aws/amazon-chime-sdk-js
# Update NScale uplink policy to scale down resolution based on number of remote videos

**Description of changes:**
Previously, NScale video uplink policy only update the bitrate when the number of remote videos increase.  
Now, we also try to scale down the video resolution based on the number of remote videos as well as the current video capture quality, except for the active speaker who would always have no resolution scale down.

The default logic would be to scale down the capture resolution to 540p, 480p, 360p, 270, and 180p when the number of videos increases to more than 2, 4, 8, 12, 16. For less than or equal to 2 videos, the resolution will not be scale down. 

Below is the table of what the bitrate and resolution would be with the default 1400 max bitrate, 540p video resolution and 15 fps:
| Number of videos  | Max bitrates (kbps) | Resolution (p) | Total max bitrates (kbps) | Total pixels |
| ------------- | ------------- | ------------- | ------------- | ------------- | 
| 1 | 1400 | 540  |  1400 | 518,400
| 2 | 1400 | 540  | 2800 | 1,036,800
| 3 | 933 | 540  | 2799 | 1,555,200
| 4 | 933 | 540  | 3732 | 2,073,600
| 5 | 746 | 480  | 3730 | 2,048,000
| 6 | 641 | 480  | 3846 | 2,457,600
| 7 | 566 | 480  | 3962 | 2,867,200
| 8 | 509 | 480  | 4072 | 3,276,800
| 9 | 466 | 360  | 4194 | 2,073,600
| 10 | 431 | 360  | 4310 | 2,304,000
| 11| 402 | 360  |  4422 | 2,534,400
| 12 | 378 | 360  | 4536 | 2,764,800
| 13 | 358 | 270  | 4654 | 1,684,800
| 14 | 340 | 270  | 4760 | 1,814,400
| 15 | 325 | 270  | 4875 | 1,944,000
| 16 | 312 | 270  | 4992 | 2,073,600
| 17 | 301 | 180  | 5117 |  979,200
| 18 | 290 | 180 | 5220 | 1,036,800
| 19 | 281 | 180  | 5339 | 1,094,400
| 20 | 273 | 180  | 5460 | 1,152,000
| 21 | 265 | 180  | 5565 | 1,209,600
| 22 | 258 | 180  | 5676 | 1,267,200
| 23 | 252 | 180  | 5796 | 1,324,800
| 24 | 246 | 180  | 5904 | 1,382,400
| 25 | 241 | 180  | 6025 | 1,440,000

**Testing**

1. Have you successfully run `npm run build:release` locally? Yes
2. How did you test these changes?
    - Have 25 attendees join a meeting with everyone muted.
    - Turn on 2 videos and make sure they are the same original capture resolution (e.g., 720p).
    - Turn on 2 more videos (4 total) and verify all 4 have 540p resolution.
    - Turn on 4 more videos and verify all 8 have 480p resolution.
    - Turn on 4 more videos and verify all 12 have 360p resolution
    - Turn on 4 more videos and verify all 16 have 270p resolution
    - Turn on 7 more videos and verify  all 25 have 180p resolution
    - Now repeat the test but with active speaker and verify that the resolution of the active speaker is not scaled down.
    Note that for Firefox the resolution may be different (for example, 540p will be 480p). 
3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it? Yes using the meeting demo.
4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved? Yes
5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved? Yes.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
